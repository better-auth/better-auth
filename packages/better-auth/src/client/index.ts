import type { BetterAuthPlugin } from "../types";
import type { BetterAuthClientPlugin } from "./types";
export * from "./vanilla";
export * from "./query";
export * from "./types";

export const InferPlugin = <T extends BetterAuthPlugin>() => {
	return {
		id: "infer-server-plugin",
		$InferServerPlugin: {} as T,
	} satisfies BetterAuthClientPlugin;
};

export type * from "@better-fetch/fetch";
//@ts-expect-error
export type * from "zod";
