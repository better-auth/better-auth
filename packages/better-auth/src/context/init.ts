import { BetterAuthError } from "@better-auth/core/error";
import { getKyselyDatabaseType } from "../adapters/kysely-adapter/dialect";
import { getAdapterWithKysely } from "../db/adapter-kysely";
import { getMigrations } from "../db/get-migration";
import type { BetterAuthOptions, DBAdapter } from "../types";
import { createAuthContext } from "./base";

const initBase = async (adapter: DBAdapter, options: BetterAuthOptions) => {
	// Get database type using Kysely's dialect detection
	const getDatabaseType = (database: BetterAuthOptions["database"]) =>
		getKyselyDatabaseType(database) || "unknown";

	// Use base context creation
	const ctx = await createAuthContext(adapter, options, getDatabaseType);

	// Add runMigrations with Kysely support
	ctx.runMigrations = async function () {
		// only run migrations if database is provided and it's not an adapter
		if (!options.database || "updateMany" in options.database) {
			throw new BetterAuthError(
				"Database is not provided or it's an adapter. Migrations are only supported with a database instance.",
			);
		}
		const { runMigrations } = await getMigrations(options);
		await runMigrations();
	};

	return ctx;
};

export const init = async (options: BetterAuthOptions) => {
	const adapter = await getAdapterWithKysely(options);
	return initBase(adapter, options);
};
