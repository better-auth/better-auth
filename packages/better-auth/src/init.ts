import { defu } from "defu";
import { hashPassword, verifyPassword } from "./crypto/password";
import { createInternalAdapter } from "./db";
import { getAuthTables } from "./db/get-tables";
import { getAdapter } from "./db/utils";
import type {
	Adapter,
	BetterAuthOptions,
	BetterAuthPlugin,
	SecondaryStorage,
	Session,
	User,
} from "./types";
import { DEFAULT_SECRET } from "./utils/constants";
import {
	type BetterAuthCookies,
	createCookieGetter,
	getCookies,
} from "./cookies";
import { createLogger } from "./utils/logger";
import { socialProviderList, socialProviders } from "./social-providers";
import type { OAuthProvider } from "./oauth2";
import { generateId } from "./utils";
import { env, isProduction } from "./utils/env";
import { checkPassword } from "./utils/password";
import { getBaseURL } from "./utils/url";

export const init = async (options: BetterAuthOptions) => {
	const adapter = await getAdapter(options);
	const plugins = options.plugins || [];
	const internalPlugins = getInternalPlugins(options);
	const logger = createLogger(options.logger);

	const baseURL = getBaseURL(options.baseURL, options.basePath);

	const secret =
		options.secret ||
		env.BETTER_AUTH_SECRET ||
		env.AUTH_SECRET ||
		DEFAULT_SECRET;

	if (secret === DEFAULT_SECRET) {
		if (isProduction) {
			logger.error(
				"You are using the default secret. Please set `BETTER_AUTH_SECRET` in your environment variables or pass `secret` in your auth config.",
			);
		}
	}

	options = {
		...options,
		secret,
		baseURL: baseURL ? new URL(baseURL).origin : "",
		basePath: options.basePath || "/api/auth",
		plugins: plugins.concat(internalPlugins),
		emailAndPassword: {
			...options.emailAndPassword,
			enabled: options.emailAndPassword?.enabled ?? false,
			autoSignIn: options.emailAndPassword?.autoSignIn ?? true,
		},
	};
	const cookies = getCookies(options);
	const tables = getAuthTables(options);
	const providers = Object.keys(options.socialProviders || {})
		.map((key) => {
			const value = options.socialProviders?.[key as "github"]!;
			if (value.enabled === false) {
				return null;
			}
			if (!value.clientId || !value.clientSecret) {
				logger.warn(
					`Social provider ${key} is missing clientId or clientSecret`,
				);
			}
			return socialProviders[key as (typeof socialProviderList)[number]](value);
		})
		.filter((x) => x !== null);

	const ctx: AuthContext = {
		appName: options.appName || "Better Auth",
		socialProviders: providers,
		options,
		tables,
		trustedOrigins: getTrustedOrigins(options),
		baseURL: baseURL || "",
		sessionConfig: {
			updateAge: options.session?.updateAge || 24 * 60 * 60, // 24 hours
			expiresIn: options.session?.expiresIn || 60 * 60 * 24 * 7, // 7 days
		},
		secret,
		rateLimit: {
			...options.rateLimit,
			enabled: options.rateLimit?.enabled ?? isProduction,
			window: options.rateLimit?.window || 10,
			max: options.rateLimit?.max || 100,
			storage:
				options.rateLimit?.storage || options.secondaryStorage
					? ("secondary-storage" as const)
					: ("memory" as const),
		},
		authCookies: cookies,
		logger: logger,
		uuid: generateId,
		session: null,
		secondaryStorage: options.secondaryStorage,
		password: {
			hash: options.emailAndPassword?.password?.hash || hashPassword,
			verify: options.emailAndPassword?.password?.verify || verifyPassword,
			config: {
				minPasswordLength: options.emailAndPassword?.minPasswordLength || 8,
				maxPasswordLength: options.emailAndPassword?.maxPasswordLength || 128,
			},
			checkPassword,
		},
		adapter: adapter,
		internalAdapter: createInternalAdapter(adapter, {
			options,
			hooks: options.databaseHooks ? [options.databaseHooks] : [],
		}),
		createAuthCookie: createCookieGetter(options),
	};
	let { context } = runPluginInit(ctx);
	return context;
};

export type AuthContext = {
	options: BetterAuthOptions;
	appName: string;
	baseURL: string;
	trustedOrigins: string[];
	session: {
		session: Session;
		user: User;
	} | null;
	socialProviders: OAuthProvider[];
	authCookies: BetterAuthCookies;
	logger: ReturnType<typeof createLogger>;
	rateLimit: {
		enabled: boolean;
		window: number;
		max: number;
		storage: "memory" | "database" | "secondary-storage";
	} & BetterAuthOptions["rateLimit"];
	adapter: Adapter;
	internalAdapter: ReturnType<typeof createInternalAdapter>;
	createAuthCookie: ReturnType<typeof createCookieGetter>;
	secret: string;
	sessionConfig: {
		updateAge: number;
		expiresIn: number;
	};
	uuid: (size?: number) => string;
	secondaryStorage: SecondaryStorage | undefined;
	password: {
		hash: (password: string) => Promise<string>;
		verify: (password: string, hash: string) => Promise<boolean>;
		config: {
			minPasswordLength: number;
			maxPasswordLength: number;
		};
		checkPassword: typeof checkPassword;
	};
	tables: ReturnType<typeof getAuthTables>;
};

function runPluginInit(ctx: AuthContext) {
	let options = ctx.options;
	const plugins = options.plugins || [];
	let context: AuthContext = ctx;
	const dbHooks: BetterAuthOptions["databaseHooks"][] = [];
	for (const plugin of plugins) {
		if (plugin.init) {
			const result = plugin.init(ctx);
			if (typeof result === "object") {
				if (result.options) {
					if (result.options.databaseHooks) {
						dbHooks.push(result.options.databaseHooks);
					}
					options = defu(options, result.options);
				}
				if (result.context) {
					context = {
						...context,
						...(result.context as Partial<AuthContext>),
					};
				}
			}
		}
	}
	// Add the global database hooks last
	dbHooks.push(options.databaseHooks);
	context.internalAdapter = createInternalAdapter(ctx.adapter, {
		options,
		hooks: dbHooks.filter((u) => u !== undefined),
	});
	context.options = options;
	return { context };
}

function getInternalPlugins(options: BetterAuthOptions) {
	const plugins: BetterAuthPlugin[] = [];
	if (options.advanced?.crossSubDomainCookies?.enabled) {
		//TODO: add internal plugin
	}
	return plugins;
}

function getTrustedOrigins(options: BetterAuthOptions) {
	const baseURL = getBaseURL(options.baseURL, options.basePath);
	if (!baseURL) {
		return [];
	}
	const trustedOrigins = [new URL(baseURL).origin];
	if (options.trustedOrigins) {
		trustedOrigins.push(...options.trustedOrigins);
	}
	const envTrustedOrigins = env.BETTER_AUTH_TRUSTED_ORIGINS;
	if (envTrustedOrigins) {
		trustedOrigins.push(...envTrustedOrigins.split(","));
	}
	return trustedOrigins;
}
