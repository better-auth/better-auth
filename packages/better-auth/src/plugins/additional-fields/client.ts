import type { NeverToUnknown } from "@prisma/client/runtime/library";
import type { FieldAttribute } from "../../db";
import type { BetterAuthClientPlugin, BetterAuthOptions } from "../../types";
import type { BetterAuthPlugin } from "../../types";

export const inferAdditionalFields = <
	T,
	S extends {
		user?: {
			[key: string]: FieldAttribute;
		};
		session?: {
			[key: string]: FieldAttribute;
		};
	} = {},
>(
	schema?: S,
) => {
	type Opts = T extends BetterAuthOptions
		? T
		: T extends {
					options: BetterAuthOptions;
				}
			? T["options"]
			: never;

	type Plugin = Opts extends never
		? S extends {
				user?: {
					[key: string]: FieldAttribute;
				};
				session?: {
					[key: string]: FieldAttribute;
				};
			}
			? {
					id: "additional-fields-client";
					schema: {
						user: {
							fields: S["user"] extends object ? S["user"] : {};
						};
						session: {
							fields: S["session"] extends object ? S["session"] : {};
						};
					};
				}
			: never
		: Opts extends BetterAuthOptions
			? {
					id: "additional-fields";
					schema: {
						user: {
							fields: Opts["user"] extends {
								additionalFields: infer U;
							}
								? U
								: {};
						};
						session: {
							fields: Opts["session"] extends {
								additionalFields: infer U;
							}
								? U
								: {};
						};
					};
				}
			: never;

	return {
		id: "additional-fields-client",
		$InferServerPlugin: {} as Plugin extends BetterAuthPlugin
			? Plugin
			: undefined,
	} satisfies BetterAuthClientPlugin;
};
