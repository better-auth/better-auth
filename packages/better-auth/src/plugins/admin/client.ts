import type { admin } from ".";
import type { BetterAuthClientPlugin } from "../../types";

export const adminClient = () => {
	return {
		id: "better-auth-client",
		$InferServerPlugin: {} as ReturnType<typeof admin>,
		pathMethods: {
			"/admin/list-users": "GET",
		},
	} satisfies BetterAuthClientPlugin;
};
