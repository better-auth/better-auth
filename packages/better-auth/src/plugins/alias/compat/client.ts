import type { BetterAuthClientPlugin } from "@better-auth/core";
import type { LiteralString } from "../../../types/helper";
import { getBaseURL } from "../../../utils/url";
import { resolveURL, toCamelCase, updateMatcher } from "../utils";
import { BetterAuthError } from "@better-auth/core/error";
import { capitalizeFirstLetter } from "../../../utils";
import type {
	InferAliasCompatClientPlugin,
	InferAliasedClientPlugin_base,
	InferClientMeta,
} from "../types/client";

export type AliasCompatClientOptions = {
	/**
	 * Additional endpoints that should be prefixed.
	 *
	 * Use this to define endpoints that should be prefixed
	 * and are not present in the plugin's pathMethods.
	 */
	includeEndpoints?: LiteralString[];
};

/**
 * Wraps a client plugin and selectively prefixes its endpoints with
 * a sub-path to preserve interoperability with an aliased plugin.
 *
 * @param aliasedPlugin - The aliased plugin indicating which parts need to be rewritten
 * @param plugin - The plugin to wrap
 * @param options - Additional configuration
 * @returns A new plugin that's compatible with the aliasedPlugin
 */
export function aliasCompatClient<
	AliasedPlugin extends BetterAuthClientPlugin,
	T extends BetterAuthClientPlugin,
	O extends AliasCompatClientOptions,
>(aliasedPlugin: AliasedPlugin, plugin: T, options?: O) {
	if (!("~meta" in aliasedPlugin)) {
		throw new BetterAuthError(
			"aliasedPlugin",
			"Invalid aliasedPlugin provided.",
		);
	}
	const {
		prefix,
		options: aliasOptions,
		signals,
	} = aliasedPlugin["~meta"] as InferClientMeta<AliasedPlugin>;
	const camelCasePrefix = toCamelCase(prefix);
	const compatPlugin: BetterAuthClientPlugin = {
		...plugin,
	};

	const includeEndpoints = [
		...new Set([
			...(options?.includeEndpoints || []),
			...(aliasOptions?.includeEndpoints || []),
			...Object.keys(aliasedPlugin.pathMethods || {})
				.filter((path) => path.startsWith(prefix))
				.map((path) => path.slice(prefix.length)),
		]),
	];

	// Update atomListeners matchers
	if (plugin.atomListeners) {
		compatPlugin.atomListeners = () => {
			const originalListeners =
				(typeof plugin.atomListeners === "function"
					? plugin.atomListeners()
					: plugin.atomListeners) || [];

			return originalListeners.map((listener) => ({
				signal:
					listener.signal !== "$sessionSignal" &&
					signals?.includes(`${listener.signal}`) &&
					aliasOptions?.prefixAtoms
						? `${listener.signal}${capitalizeFirstLetter(camelCasePrefix)}`
						: listener.signal,
				matcher: updateMatcher({
					matcher: listener.matcher,
					prefix,
					excludeEndpoints: aliasOptions?.excludeEndpoints,
				}),
			}));
		};
	}

	// Wrap getActions to prefix specific path-based action
	if (plugin.getActions) {
		compatPlugin.getActions = ($fetch, store, clientOptions) => {
			const baseURL = getBaseURL(
				clientOptions?.baseURL,
				clientOptions?.basePath,
				undefined,
			);
			return plugin.getActions!(
				((url: string | URL, ...opts: any[]) => {
					return $fetch(
						resolveURL(
							{
								url,
								baseURL,
							},
							includeEndpoints,
							prefix,
							"include",
						),
						...opts,
					);
				}) as any,
				store,
				clientOptions,
			);
		};
	}

	// Wrap getAtoms to prefix specific path-based action
	if (plugin.getAtoms) {
		compatPlugin.getAtoms = ($fetch, clientOptions) => {
			const baseURL = getBaseURL(
				clientOptions?.baseURL,
				clientOptions?.basePath,
				undefined,
			);
			return plugin.getAtoms!(
				((url: string | URL, ...opts: any[]) => {
					return $fetch(
						resolveURL({ url, baseURL }, includeEndpoints, prefix, "include"),
						...opts,
					);
				}) as any,
				clientOptions,
			);
		};
	}

	// Update fetchPlugins hooks to prefix specific paths
	if (plugin.fetchPlugins) {
		compatPlugin.fetchPlugins = plugin.fetchPlugins.map((fetchPlugin) => ({
			...fetchPlugin,
			init(url, options) {
				const resolvedURL = resolveURL(
					{
						url,
						baseURL: options?.baseURL,
					},
					includeEndpoints,
					prefix,
					"include",
				);

				return fetchPlugin.init
					? fetchPlugin.init?.(resolvedURL, options)
					: {
							url: resolvedURL,
							options,
						};
			},
		}));
	}

	return compatPlugin as InferAliasCompatClientPlugin<
		InferAliasedClientPlugin_base<
			InferClientMeta<AliasedPlugin>["prefix"],
			AliasedPlugin,
			InferClientMeta<AliasedPlugin>["options"]
		>,
		T,
		O
	>;
}
