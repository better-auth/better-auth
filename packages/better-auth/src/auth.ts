import type { Endpoint, Prettify } from "better-call";
import { getEndpoints, router } from "./api";
import { init } from "./init";
import type { BetterAuthOptions } from "./types/options";
import type { InferPluginTypes, InferSession, InferUser } from "./types";
import { getBaseURL } from "./utils/url";

type InferAPI<API> = Omit<
	API,
	API extends { [key in infer K]: Endpoint }
		? K extends string
			? API[K]["options"]["metadata"] extends { isAction: false }
				? K
				: never
			: never
		: never
>;

export const betterAuth = <O extends BetterAuthOptions>(options: O) => {
	const authContext = init(options);
	const { api } = getEndpoints(authContext, options);

	return {
		handler: async (request: Request) => {
			const ctx = await authContext;
			const basePath = ctx.options.basePath || "/api/auth";
			const url = new URL(request.url);
			if (!ctx.options.baseURL) {
				const baseURL =
					getBaseURL(undefined, basePath) || `${url.origin}${basePath}`;
				ctx.options.baseURL = baseURL;
				ctx.baseURL = baseURL;
			}
			ctx.trustedOrigins = [url.origin, ...(ctx.options.trustedOrigins || [])];
			if (!ctx.options.baseURL) {
				return new Response("Base URL not set", { status: 400 });
			}
			if (url.pathname === basePath || url.pathname === `${basePath}/`) {
				return new Response("Welcome to BetterAuth", { status: 200 });
			}
			const { handler } = router(ctx, options);
			return handler(request);
		},
		api: api as InferAPI<typeof api>,
		options: options as O,
		$Infer: {} as {
			Session: {
				session: Prettify<InferSession<O>>;
				user: Prettify<InferUser<O>>;
			};
		} & InferPluginTypes<O>,
	};
};

export type Auth = {
	handler: (request: Request) => Promise<Response>;
	api: InferAPI<ReturnType<typeof router>["endpoints"]>;
	options: BetterAuthOptions;
};
