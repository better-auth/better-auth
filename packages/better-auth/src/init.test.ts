import { afterAll, describe, expect, it, vi } from "vitest";
import { init } from "./init";
import Database from "better-sqlite3";
import { betterAuth } from "./auth";
import { createAuthClient } from "./client";
import { getTestInstance } from "./test-utils/test-instance";

describe("init", async () => {
	const database = new Database(":memory:");

	it("should match config", async () => {
		const res = await init({
			baseURL: "http://localhost:3000",
			database,
		});
		expect(res).toMatchSnapshot();
	});

	it("should infer BASE_URL from env", async () => {
		vi.stubEnv("BETTER_AUTH_URL", "http://localhost:5147");
		const res = await init({
			database,
		});
		expect(res.options.baseURL).toBe("http://localhost:5147");
		expect(res.baseURL).toBe("http://localhost:5147/api/auth");
		vi.unstubAllEnvs();
	});

	it("should respect base path", async () => {
		const res = await init({
			database,
			basePath: "/custom-path",
			baseURL: "http://localhost:5147",
		});
		expect(res.baseURL).toBe("http://localhost:5147/custom-path");
	});

	it("should work with base path", async () => {
		const { client } = await getTestInstance({
			basePath: "/custom-path",
		});

		await client.$fetch("/ok", {
			onSuccess: (ctx) => {
				expect(ctx.data).toMatchObject({
					ok: true,
				});
			},
		});
	});

	it("should execute plugins init", async () => {
		const newBaseURL = "http://test.test";
		const res = await init({
			baseURL: "http://localhost:3000",
			database,
			plugins: [
				{
					id: "test",
					init: () => {
						return {
							context: {
								baseURL: newBaseURL,
							},
						};
					},
				},
			],
		});
		expect(res.baseURL).toBe(newBaseURL);
	});

	it("should work with custom path", async () => {
		const customPath = "/custom-path";
		const ctx = await init({
			database,
			basePath: customPath,
			baseURL: "http://localhost:3000",
		});
		expect(ctx.baseURL).toBe(`http://localhost:3000${customPath}`);

		const res = betterAuth({
			baseURL: "http://localhost:3000",
			database,
			basePath: customPath,
		});

		const client = createAuthClient({
			baseURL: `http://localhost:3000/custom-path`,
			fetchOptions: {
				customFetchImpl: async (url, init) => {
					return res.handler(new Request(url, init));
				},
			},
		});
		const ok = await client.$fetch("/ok");
		expect(ok.data).toMatchObject({
			ok: true,
		});
	});
});
