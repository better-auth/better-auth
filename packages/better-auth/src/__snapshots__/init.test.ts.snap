// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`init > should match config 1`] = `
{
  "adapter": {
    "create": [Function],
    "delete": [Function],
    "deleteMany": [Function],
    "findMany": [Function],
    "findOne": [Function],
    "id": "kysely",
    "options": {
      "type": "sqlite",
    },
    "update": [Function],
    "updateMany": [Function],
  },
  "appName": "Better Auth",
  "authCookies": {
    "dontRememberToken": {
      "name": "better-auth.dont_remember",
      "options": {
        "httpOnly": true,
        "path": "/",
        "sameSite": "lax",
        "secure": false,
      },
    },
    "sessionData": {
      "name": "better-auth.session_data",
      "options": {
        "httpOnly": true,
        "maxAge": 300,
        "path": "/",
        "sameSite": "lax",
        "secure": false,
      },
    },
    "sessionToken": {
      "name": "better-auth.session_token",
      "options": {
        "httpOnly": true,
        "maxAge": 604800,
        "path": "/",
        "sameSite": "lax",
        "secure": false,
      },
    },
  },
  "baseURL": "http://localhost:3000/api/auth",
  "createAuthCookie": [Function],
  "generateId": [Function],
  "internalAdapter": {
    "createAccount": [Function],
    "createOAuthUser": [Function],
    "createSession": [Function],
    "createUser": [Function],
    "createVerificationValue": [Function],
    "deleteSession": [Function],
    "deleteSessions": [Function],
    "deleteUser": [Function],
    "deleteVerificationByIdentifier": [Function],
    "deleteVerificationValue": [Function],
    "findAccount": [Function],
    "findAccountByUserId": [Function],
    "findAccounts": [Function],
    "findSession": [Function],
    "findSessions": [Function],
    "findUserByEmail": [Function],
    "findUserById": [Function],
    "findVerificationValue": [Function],
    "linkAccount": [Function],
    "listSessions": [Function],
    "listUsers": [Function],
    "updateAccount": [Function],
    "updatePassword": [Function],
    "updateSession": [Function],
    "updateUser": [Function],
    "updateUserByEmail": [Function],
    "updateVerificationValue": [Function],
  },
  "logger": {
    "debug": [Function],
    "error": [Function],
    "info": [Function],
    "success": [Function],
    "warn": [Function],
  },
  "options": {
    "basePath": "/api/auth",
    "baseURL": "http://localhost:3000",
    "database": Database {
      "inTransaction": false,
      "memory": true,
      "name": ":memory:",
      "open": true,
      "readonly": false,
    },
    "emailAndPassword": {
      "autoSignIn": true,
      "enabled": false,
    },
    "plugins": [],
    "secret": "better-auth-secret-123456789",
  },
  "password": {
    "checkPassword": [Function],
    "config": {
      "maxPasswordLength": 128,
      "minPasswordLength": 8,
    },
    "hash": [Function],
    "verify": [Function],
  },
  "rateLimit": {
    "enabled": false,
    "max": 100,
    "storage": "memory",
    "window": 10,
  },
  "secondaryStorage": undefined,
  "secret": "better-auth-secret-123456789",
  "session": null,
  "sessionConfig": {
    "expiresIn": 604800,
    "updateAge": 86400,
  },
  "socialProviders": [],
  "tables": {
    "account": {
      "fields": {
        "accessToken": {
          "fieldName": "accessToken",
          "required": false,
          "type": "string",
        },
        "accountId": {
          "fieldName": "accountId",
          "required": true,
          "type": "string",
        },
        "createdAt": {
          "fieldName": "createdAt",
          "required": true,
          "type": "date",
        },
        "expiresAt": {
          "fieldName": "expiresAt",
          "required": false,
          "type": "date",
        },
        "idToken": {
          "fieldName": "idToken",
          "required": false,
          "type": "string",
        },
        "password": {
          "fieldName": "password",
          "required": false,
          "type": "string",
        },
        "providerId": {
          "fieldName": "providerId",
          "required": true,
          "type": "string",
        },
        "refreshToken": {
          "fieldName": "refreshToken",
          "required": false,
          "type": "string",
        },
        "updatedAt": {
          "fieldName": "updatedAt",
          "required": true,
          "type": "date",
        },
        "userId": {
          "fieldName": "userId",
          "references": {
            "field": "id",
            "model": "user",
            "onDelete": "cascade",
          },
          "required": true,
          "type": "string",
        },
      },
      "modelName": "account",
      "order": 3,
    },
    "session": {
      "fields": {
        "createdAt": {
          "fieldName": "createdAt",
          "required": true,
          "type": "date",
        },
        "expiresAt": {
          "fieldName": "expiresAt",
          "required": true,
          "type": "date",
        },
        "ipAddress": {
          "fieldName": "ipAddress",
          "required": false,
          "type": "string",
        },
        "token": {
          "fieldName": "token",
          "required": true,
          "type": "string",
          "unique": true,
        },
        "updatedAt": {
          "fieldName": "updatedAt",
          "required": true,
          "type": "date",
        },
        "userAgent": {
          "fieldName": "userAgent",
          "required": false,
          "type": "string",
        },
        "userId": {
          "fieldName": "userId",
          "references": {
            "field": "id",
            "model": "user",
            "onDelete": "cascade",
          },
          "required": true,
          "type": "string",
        },
      },
      "modelName": "session",
      "order": 2,
    },
    "user": {
      "fields": {
        "createdAt": {
          "defaultValue": [Function],
          "fieldName": "createdAt",
          "required": true,
          "type": "date",
        },
        "email": {
          "fieldName": "email",
          "required": true,
          "type": "string",
          "unique": true,
        },
        "emailVerified": {
          "defaultValue": [Function],
          "fieldName": "emailVerified",
          "required": true,
          "type": "boolean",
        },
        "image": {
          "fieldName": "image",
          "required": false,
          "type": "string",
        },
        "name": {
          "fieldName": "name",
          "required": true,
          "type": "string",
        },
        "updatedAt": {
          "defaultValue": [Function],
          "fieldName": "updatedAt",
          "required": true,
          "type": "date",
        },
      },
      "modelName": "user",
      "order": 1,
    },
    "verification": {
      "fields": {
        "createdAt": {
          "defaultValue": [Function],
          "fieldName": "createdAt",
          "required": false,
          "type": "date",
        },
        "expiresAt": {
          "fieldName": "expiresAt",
          "required": true,
          "type": "date",
        },
        "identifier": {
          "fieldName": "identifier",
          "required": true,
          "type": "string",
        },
        "updatedAt": {
          "defaultValue": [Function],
          "fieldName": "updatedAt",
          "required": false,
          "type": "date",
        },
        "value": {
          "fieldName": "value",
          "required": true,
          "type": "string",
        },
      },
      "modelName": "verification",
      "order": 4,
    },
  },
  "trustedOrigins": [
    "http://localhost:3000",
  ],
}
`;
